src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,498 @@
+							hostName, nodePort, entry->reservedConnectionCount)));
+}
+
+
+/*
+ * EnsurePrimaryNodesHaveReservedConnection is a wrapper around
+ * ReserveSharedConnectionCounterForNodeListIfNeeded.
+ */
+void
+EnsurePrimaryNodesHaveReservedConnection(int count)
+{
+	/*
+	 * Prevent addition of new nodes. When WritableStandbyCoordinator is true, we
+	 * don't need to acquire the lock because node addition is not possible
+	 * anyway.
+	 */
+	LOCKMODE lockMode =
+		(RecoveryInProgress() && WritableStandbyCoordinator) ? NoLock : ShareLock;
+	List *primaryNodeList = ActivePrimaryNodeList(lockMode);
+
+	/*
+	 * We need to reserve connection one by one because the underlying
+	 * functions (e.g., WaitLoopForSharedConnection) do not implement
+	 * reserving multiple connections.
+	 */
+	int reserveIndex = 0;
+	for (; reserveIndex < count; ++reserveIndex)
+	{
+		ReserveSharedConnectionCounterForNodeListIfNeeded(primaryNodeList);
+	}
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForNodeListIfNeeded reserves a shared connection
+ * counter per node in the nodeList unless:
+ *  - there is at least one connection to the node
+ */
+static void
+ReserveSharedConnectionCounterForNodeListIfNeeded(List *nodeList)
+{
+	if (!IsReservationRequired())
+	{
+		return;
+	}
+
+	/*
+	 * We sort the workerList because adaptive connection management
+	 * (e.g., OPTIONAL_CONNECTION) requires any concurrent executions
+	 * to wait for the connections in the same order to prevent any
+	 * starvation. If we don't sort, we might end up with:
+	 *      Execution 1: Get connection for worker 1, wait for worker 2
+	 *      Execution 2: Get connection for worker 2, wait for worker 1
+	 *
+	 *  and, none could proceed. Instead, we enforce every execution establish
+	 *  the required connections to workers in the same order.
+	 */
+	nodeList = SortList(nodeList, CompareWorkerNodes);
+
+	char *databaseName = get_database_name(MyDatabaseId);
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (ConnectionAvailableToNode(workerNode->workerName, workerNode->workerPort,
+									  CurrentUserName(), databaseName))
+		{
+			/*
+			 * The same user has already an active connection for the node. It
+			 * means that the execution can use the same connection, so reservation
+			 * is not necessary.
+			 */
+			continue;
+		}
+
+		/*
+		 * We are trying to be defensive here by ensuring that the required hash
+		 * table entry can be allocated. The main goal is that we don't want to be
+		 * in a situation where shared connection counter is incremented but not
+		 * the local reserved counter due to out-of-memory.
+		 *
+		 * Note that shared connection stats operate on the shared memory, and we
+		 * pre-allocate all the necessary memory. In other words, it would never
+		 * throw out of memory error.
+		 */
+		ReservedConnectionCounterHashEntry *hashEntry =
+			AllocateOrGetReservedConectionEntry(workerNode->workerName,
```

Comment:
As discussed:

We should check whether a hash entry exists here and `continue;` if it does. That way, we prevent any nested/future callers from making additional reservations.

We could use the found flag in hash_search to avoid adding an additional HASH_FIND lookup.

Reply:
I guess, with this we don't need the `ConnectionAvailableToNode()` check as well. Because, if we have an entry, it is guaranteed that the connection is avaliable for this command as well, right?

(consider userName is added to the hash)